homeassistant/components/unifiprotect/select.py

```diff
@@ -93,44 +86,131 @@
 
 
 @dataclass
-class ProtectSelectEntityDescription(ProtectRequiredKeysMixin, SelectEntityDescription):
+class ProtectSelectEntityDescription(ProtectSetableKeysMixin, SelectEntityDescription):
     """Describes UniFi Protect Select entity."""
 
     ufp_options: list[dict[str, Any]] | None = None
+    ufp_options_callable: Callable[
+        [ProtectApiClient], list[dict[str, Any]]
+    ] | None = None
     ufp_enum_type: type[Enum] | None = None
-    ufp_set_function: str | None = None
+    ufp_set_method: str | None = None
+
+
+def _get_viewer_options(api: ProtectApiClient) -> list[dict[str, Any]]:
+    return [
+        {"id": item.id, "name": item.name} for item in api.bootstrap.liveviews.values()
+    ]
+
+
+def _get_doorbell_options(api: ProtectApiClient) -> list[dict[str, Any]]:
+    default_message = api.bootstrap.nvr.doorbell_settings.default_message_text
+    messages = api.bootstrap.nvr.doorbell_settings.all_messages
+    built_messages = ({"id": item.type.value, "name": item.text} for item in messages)
+
+    return [
+        {"id": "", "name": f"Default Message ({default_message})"},
+        *built_messages,
+    ]
+
+
+def _get_paired_camera_options(api: ProtectApiClient) -> list[dict[str, Any]]:
+    options = [{"id": TYPE_EMPTY_VALUE, "name": "Not Paired"}]
+    for camera in api.bootstrap.cameras.values():
+        options.append({"id": camera.id, "name": camera.name})
+
+    return options
+
+
+def _get_viewer_current(obj: Any) -> str:
+    assert isinstance(obj, Viewer)
+    return obj.liveview_id
+
+
+def _get_light_motion_current(obj: Any) -> str:
+    assert isinstance(obj, Light)
+    # a bit of extra to allow On Motion Always/Dark
+    if (
+        obj.light_mode_settings.mode == LightModeType.MOTION
+        and obj.light_mode_settings.enable_at == LightModeEnableType.DARK
+    ):
+        return f"{LightModeType.MOTION.value}Dark"
+    return obj.light_mode_settings.mode.value
+
+
+def _get_doorbell_current(obj: Any) -> str | None:
+    assert isinstance(obj, Camera)
+    if obj.lcd_message is None:
+        return None
+    return obj.lcd_message.text
+
+
+async def _set_light_mode(obj: Any, mode: str) -> None:
+    assert isinstance(obj, Light)
+    lightmode, timing = LIGHT_MODE_TO_SETTINGS[mode]
+    await obj.set_light_settings(
+        LightModeType(lightmode),
+        enable_at=None if timing is None else LightModeEnableType(timing),
+    )
+
+
+async def _set_paired_camera(obj: Any, camera_id: str) -> None:
+    assert isinstance(obj, Light)
+    if camera_id == TYPE_EMPTY_VALUE:
+        camera: Camera | None = None
+    else:
+        camera = obj.api.bootstrap.cameras.get(camera_id)
+    await obj.set_paired_camera(camera)
+
+
+async def _set_doorbell_message(obj: Any, message: str) -> None:
```

Comment:
Why not type `obj` as `Camera` if that's required?

Reply:
mypy complains otherwise because the function signature does not match. It is always Camera though. So, either an assert or a type ignore is needed.

Reply:
Maybe something like this
https://github.com/bdraco/home-assistant/commit/2a3a6b48b24b1b6db5b9031665659d060448906c

Reply:
That is more exact but does not really provide any different benefit. Still have to assert the type or ignore the assignment for mypy.

Reply:
Looking back at each of them, they are each the main device types (NVR/Camera/Light/Sensor/Viewer). So, I could different EntityDescriptions for each one and that might make all of the linting work nicely. 

Since this PR is already pretty big, I think I could do another follow up PR to try to reduce/remove any asserts/comments for entity descriptions specifically.

Reply:
This PR is already big enough so a follow up would be good

Reply:
Update here. It looks like my idea will work, but it will still require a type ignore on the child EntityDescription class:

```
homeassistant/components/unifiprotect/number.py:39: error: Definition of "ufp_set_method_fn" in base class "LightSetableKeysMixin" is incompatible with definition in base class "ProtectSetableKeysMixin"  [misc]
homeassistant/components/unifiprotect/number.py:39: error: Definition of "ufp_value_fn" in base class "LightRequiredKeysMixin" is incompatible with definition in base class "ProtectRequiredKeysMixin"  [misc]
Found 2 errors in 1 file (checked 2 source files)
```

It still greatly reduces the overrides though.

Reply:
Scratch that. If I just use `Any` on the base mixins, all of the mypy errors go away and there is no need for asserts. 

That is kind of annoying that `Any` works but the base class that they _all_ share did not. I guess I learn more about mypy every day.